### ISSUE TITLE
Pointer Aliases

### ISSUE THREAD

**Issue by TwoLaid** [hyrise/hyrise#403]
As discussed in the last fortnightly, it might be a good idea to have aliases for some common smart pointers. So something like: `using TablePtr = std::shared_ptr<Table>;`

Now, there are some details that need to be discussed. So for example, since we want smart pointers to be passed as const refs, one of these might make more sense:

```c++
using TablePtr = std::shared_ptr<Table>;
using TablePtr = const std::shared_ptr<Table>;
using TablePtr = const std::shared_ptr<Table>&;
```

Another "problem" is that we often have smart pointers to `const` objects, like `std::shared_ptr<const Table>`. How do we want to alias those?

**Comment by TwoLaid** [comment]
I'm sure @mrzzzrm and @mjendruk have opinions on this :)

**Comment by mrks** [comment]
Generally in favor, as long as it improves readability without introducing any ambiguity.

Seeing the three examples above and thinking about #398, I wonder if those aliases would make it more difficult for people to identify cases that are now frowned upon. "Don't put a `shared_ptr` into a method parameter if it doesn't have `const` before and `&` after" might be easier to check than "Don't have something in a method parameter that ends in `Ptr` if there is no `const` before or no `&` after".

**Comment by mjendruk** [comment]
We could have something like this
> "Don't put a shared_ptr into a method parameter if it doesn't have const before and & after" might be easier to check than "Don't have something in a method parameter that ends in Ptr if there is no const before or no & after".

I think we should assume that people know what a typedef is.

```C++
class Table;

using TableSPtr = std::shared<Table>;
using TableSPtr = std::shared<const Table>;
```

I would not put qualifiers such as `const` and references into typedefs. 

We could also just create shorthands for pointers:

```C++
typename <class T>
using sptr = std::shared_ptr<T>;

typename <class T>
using uptr = std::unique_ptr<T>;

// Usage
void method(const sptr<BaseAttributeVector>& vec);  // VS
void method(const std::shared_ptr<BaseAttributeVector>& table);
```

**Comment by mrks** [comment]
The first example highlights potential for confusion: If we want shared pointers to const Tables, what would their name be? People shouldn't have to remember if a `TableSPtr` points to a const table or not. Even less so, if this is completely different for e.g., a `TransactionsContextSPtr`.

What are we hoping to achieve here?

**Comment by mjendruk** [comment]
Correction: The third line in the first code block should have been
```c++
using TableSCPtr = std::shared_ptr<const Table>;
```

I agree that if we added such typedefs there should be rules, for example that `*SPtr` is always a non-const shared pointer. The aim was to shorten parameter types, since 
`const std::shared_ptr<BaseAttributeVector>& vec` is quite long when what would actually be required is a simple `const BaseAttributeVector& vec`. So as a comprise I suggested typedefs.

**Comment by mrzzzrm** [comment]
Did we conclude that a macro #define CREATE_PTR_ALIASES(class_name) that creates aliases for a bunch of common ptr types is a good idea. I like it.
@mjendruk @mrks

**Comment by mjendruk** [comment]
We didn’t decide on the naming of the actual aliases … In some files, I also saw ClassName::Ptr, we could implement this by inherting from some templated base type.

```c++
template <typename Derived>
struct PointerBase {
  using ptr = std::shared_ptr<Derived>;
  using cptr = std::shared_ptr<const Derived>;
};

**Comment by mrzzzrm** [comment]
Na, that looks weird: `PredicateNode::ptr`, I prefer PredicateNodePtr, generated by a macro. Which aliases do we want?
I want these:
 - *(S)Ptr = shared_ptr<....>
 - *C(S)Ptr = shared_ptr<const ... >
- C...(S)Ptr = const shared_ptr<...>
- C...C(S)Ptr = const shared_ptr<const ...>

**Comment by mrzzzrm** [comment]
I'd add the 'S', even though we don't have an alias for unique_ptr.
Dying to create the regex that refactors this in one go.

**Comment by mrks** [comment]
Works for me. We should add that regex to our lint.sh to make sure that we stay consistent.

**Comment by mjendruk** [comment]
I wouldn’t add aliases for `const shared_ptr<...>`. I think it makes more sense to write `const MyClassPtr`

**Comment by mrzzzrm** [comment]
We could save 5 chars, Max!!!! I'm serious about this.

**Comment by mjendruk** [comment]
No, I think it’s good to still see the common `const something&` pattern. `somethingcptr&` and it looks like it’s a non-const reference.

**Comment by mrzzzrm** [comment]
You're in revolutions way.

**Comment by mjendruk** [comment]
P.S.: You should probably wait until all the BYOD PRs are merged.
